### PR TITLE
Maintainer's guide for Homebrew on Linux (linuxbrew-core)

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] ~Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).~
- [x] ~Have you successfully run `brew style` with your changes locally?~
- [x] ~Have you successfully run `brew tests` with your changes locally?~

-----

- There are a lot more manual steps in Homebrew on Linux (Linuxbrew)
  land than there are in Homebrew (core), it feels like.
- This is a first attempt at a maintainer's guide. I have basically
  tried to brain-dump everything that Michka and Shaun have taught me
  over the months into something consumable for future Homebrew on Linux
  maintainers!
- Some of it is overly verbose - the merging docs are taken from
  [Linuxbrew/homebrew-developer's docs](https://github.com/Linuxbrew/homebrew-developer/blob/master/docs/merge-homebrew-core.md)
  which were recently updated to reflect reality but probably don't need
  to have so many words: just some instructions. :-)
- Still to document:
  - [ ] the fact that CI is broken on Homebrew/homebrew-core merge PRs
  - [ ] how to handle users requesting bottles be built for formulae that don't have them
  - [ ] how to pull bottles (`brew pull --bottle` and `brew squash-bottle-pr`)
  - [ ] merging PRs on GitHub, when do we do that?
  - [ ] more that I will doubtless think of later...

### CODE REVIEW

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
+# Homebrew on Linux Maintainer Guide
```
Comment (MikeMcQuaid) [inline docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: Maybe put `linuxbrew-core` in the title

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+### Bottling errors
+
+## Handling `brew bump-formula-pr` PRs
+
+### Formulae that exist in Homebrew/homebrew-core
+
+The `brew bump-formula-pr` command will raise PRs against the Linux
+formula repo for upstream Mac formulae when running on Linux. This
+isn't how version bumps are done for _most_ formulae. Redirect users
+to raise their PRs in Homebrew/homebrew-core:
```
Comment (MikeMcQuaid) [inline docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: Could link to https://github.com/Homebrew/brew/issues/6341 here to nudge people to help?

**Comment by MikeMcQuaid** [comment]
> Still to document:
> * [ ]  the fact that CI is broken on Homebrew/homebrew-core merge PRs
> * [ ]  how to handle users requesting bottles be built for formulae that don't have them
> * [ ]  how to pull bottles (`brew pull --bottle` and `brew squash-bottle-pr`)
> * [ ]  merging PRs on GitHub, when do we do that?
> * [ ]  more that I will doubtless think of later...

I think you can feel free to merge this before those are done, if desired. It's sometimes more motivating that way and also easier to review.

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+- it doesn't need a bottle
+- it already has a bottle
+- the formula depends on macOS to build
+- the formula's tap is Homebrew/homebrew-core (the upstream macOS repo)
+- there is already an open PR for the formula's bottle
+- the current branch is not master
+
+By default, running this script won't output the errors. To see the
+errors, run `brew find-formulae-to-bottle --verbose` separate to the
+`for` loop above.
```
[the remark this answers is not in the document]
Comment (issyl0) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: I'll happily review a PR if you find a way to do it This conversation happened in https://github.com/Linuxbrew/homebrew-developer/pull/112 too (my rationale for `--verbose` was [this](https://github.com/Linuxbrew/homebrew-developer/pull/112/commits/ac71dc990366ead4e0db2d0b7146bb39dfff5d07) and [this](https://github.com/Linuxbrew/homebrew-developer/pull/112#issuecomment-525616602)).

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+After merging changes, we must rebuild bottles for all the PRs that
+had conflicts.
+
+To do this, tap `Linuxbrew/homebrew-developer` and run the following
+command where the merge commit is `HEAD`:
+
+```sh
+for formula in brew find-formulae-to-bottle; do
+  brew build-bottle-pr --remote=<your-remote> $formula
```
[the remark this answers is not in the document]
Comment (issyl0) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: Not if we do [the GitHub Action](https://github.com/Homebrew/linuxbrew-core/pull/14838), as then that would push to `origin` and enable every maintainer to delete the branches. Personally I'm in favour of normalising pushing to `origin` on Homebrew/linuxbrew-core for the "deleting branches" reason, especially now we can get GitHub to do it for us). But that's a wider conversation outside the scope of this PR.

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+automatic updates while we do the merge:
+
+```bash
+export HOMEBREW_DEVELOPER=1
+export HOMEBREW_NO_AUTO_UPDATE=1
+```
+
+Once we've done that, we need to get access to the `merge-homebrew`
+command that will be used for the merge. To do that we have to tap
+`linuxbrew/homebrew-developer` repo:
```
[the remark this answers is not in the document]
Comment (issyl0) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]:
```suggestion
the `linuxbrew/homebrew-developer` repo:
```

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+Remote names `origin` and `homebrew` are hard-coded in
+`merge-homebrew`, while the remote pointing to your fork can have any
+name. We will call this remote `your-fork`. It will be used to submit
+a pull request for the merge.
+
+```bash
+brew install hub
+cd $(brew --repo homebrew/core)
+hub remote add homebrew
+hub remote add your-fork
```
[the remark this answers is not in the document]
Comment (issyl0) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: `hub fork` adds a remote if it doesn't exist already, without re-forking the repo if it's already forked. This is how I do things between laptops.

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjackman) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: The GitHub action can use `--remote=origin`. For interactive use, the default ought to be `--remote=$GITHUB_USER`

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjackman) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]: `opoo` writes the warnings to `$stderr`.
See https://github.com/Linuxbrew/homebrew-developer/pull/112#issuecomment-523287506

**Comment by issyl0** [comment]
Thanks, @MikeMcQuaid! Merging this as-is would be good, then we can make incremental improvements and document new stuff as we do it. :-)

File: docs/Homebrew-linuxbrew-core-Maintainer-Guide.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+### Linux-only formulae
+
+If the formula is a Linux-only formula, it either:
+- will contain the line "tag "linuxbrew""
+- won't have macOS bottles
+
+These formulae are fine for users to bump with `brew bump-formula-pr`,
+but you should request that they remove the existing `x86_64_linux`
+bottle SHA line so that CI will build a bottle for the new version
+correctly.
```
[the remark this answers is not in the document]
Comment (issyl0) [reply docs/Homebrew-linuxbrew-core-Maintainer-Guide.md]:
```suggestion
correctly. If the bottle SHA isn't removed, CI will fail with the following error:
> `--keep-old` was passed but there are changes in `sha256 => x86_64_linux`
```

**Comment by issyl0** [comment]
Bah, I missed @jonchang's last comment. Amended!

**Pull request merged**